network-mux/test/Test/Mux/ReqResp.hs

```diff
@@ -0,0 +1,202 @@
+      recv >>= liftST . k . fmap LBS.toStrict >>= go Nothing
+    go (Just bs) (CBOR.Partial k) =
+      liftST (k (Just bs)) >>= go Nothing
+    go _ (CBOR.Done trailing' _ a) | BS.null trailing' = return (Right (a, Nothing))
+                                   | otherwise         = return (Right (a, Just $ LBS.fromStrict trailing'))
+    go _ (CBOR.Fail _ _ failure) = return $ Left failure
+
+
+
+-- | Run a client using a byte 'Channel'.
+--
+runClient :: forall req resp m a.
+             ( Monad m
+             , MonadST m
+             , Serialise req
+             , Serialise resp
+             , Show req
+             , Show resp
+             )
+          => Tracer m (TraceSendRecv (MsgReqResp req resp))
+          -> Channel m
+          -> ReqRespClient req resp m a
+          -> m a
+
+runClient tracer channel@Channel {send} =
+    go Nothing
+  where
+    go :: Maybe LBS.ByteString
+       -> ReqRespClient req resp m a
+       -> m a
+    go trailing (SendMsgReq req mnext) = do
+      let msg :: MsgReqResp req resp
+          msg = MsgReq req
+      traceWith tracer (TraceSend msg)
+      send $ serialise msg
+
+      res <- withLiftST $ \liftST -> runDecoderWithChannel liftST channel trailing decode
+
+      case res of
+        Left err -> error $ "runClient: deserialise error: " ++ show err
+
+        Right (msg'@(MsgResp resp), trailing') -> do
+          traceWith tracer (TraceRecv msg')
+          mnext resp >>= go trailing'
+
+        Right (msg', _) -> error $ "runClient: wrong message " ++ show msg'
+
+    go _trailing (SendMsgDone ma) = do
+        let msg :: MsgReqResp req resp
+            msg = MsgDone
+        traceWith tracer (TraceSend msg)
+        send (serialise msg)
+        ma
+
+
+-- | Server which receives 'req' and responds with 'resp'.
+--
+data ReqRespServer req resp m a = ReqRespServer {
+    -- | The client sent us a ping message. We have no choices here, and
+    -- the response is nullary, all we have are local effects.
+    recvMsgReq  :: req -> m (resp, ReqRespServer req resp m a)
+
+    -- | The client terminated. Here we have a pure return value, but we
+    -- could have done another action in 'm' if we wanted to.
+  , recvMsgDone :: m a
+  }
+
+
+runServer :: forall req resp m a.
+             ( Monad m
+             , Serialise req
+             , Serialise resp
+             , Show req
+             , Show resp
+             , m ~ IO
+             )
+          => Tracer m (TraceSendRecv (MsgReqResp req resp))
+          -> Channel m
+          -> ReqRespServer req resp m a
+          -> m a
+
+runServer tracer channel@Channel {send} =
+    go Nothing
+  where
+    go :: Maybe LBS.ByteString
+       -> ReqRespServer req resp m a
+       -> m a
+    go trailing ReqRespServer {recvMsgReq, recvMsgDone} = do
+      res <- withLiftST $ \liftST -> runDecoderWithChannel liftST channel trailing decode
+
+      case res of
+        Left err -> error $ "runServer: deserialise error " ++ show err
```

Comment:
Would be good if we ran `traceWith` for the deserialise error too.

Reply:
Good point, I all add it.

Reply:
@karknu there's no way to do that. The reason is that `TraceSendRecv` now only allows to put trace messages not decoding errors, but I think we should extend it to something like:
```
data TraceSendRecv msg st failure
  = TraceSend msg
  | TraceRecv msg
  | TraceDecodingError st failure
  deriving Show
```
The `TraceDecodingError` could trace the current protocol state & the codec failure.  But this is out side of this PR.

Reply:
Actually I am wrong, this `TraceSendRecv` is private here, so we can change it as we see it fit in this PR.

Reply:
But extending the original trace is also a good idea.

Reply:
Then I think you should, and we can file an issue to add it to `typed-protocol` too.

Reply:
Done, yes adding an issue is a good idea.